### ISSUE TITLE
set_path_variable api enhancement

### ISSUE THREAD

**Issue by SGeeversAtVortech** [Deltares/rtc-tools#1023]
In GitLab by @jvande42b on May 6, 2018, 24:14

A very common pattern in RTC implementations is to define a new MX symbol and set its upper and lower bound equal to a timeseries. This allows users to programmatically formulate goal functions and constraints using symbols with values/bounds calculated at run time. These symbols are not part of the model, and so do not belong as inputs or variables in the modelica file.

A really valuable improvement would be to expand the api to make this a one-step procedure, e.g. 

```python
bound_ts = Timeseries(self.times(), np.full_like(self.times(), 10.0))

# Set with equality constraint (either timeseries or single value):
self.set_path_variable("var_name", bound_ts) # recognizes timeseries and uses bound_ts as upper and lower bound
self.set_path_variable("var_name", 10.0) # uses 10.0 as upper and lower bound

# Set with upper and lower bounds (either timeseries or single value):
self.set_path_variable("var_name", (bound_ts, 10.0)) # recognizes tuple and uses it for bounds
self.set_path_variable("var_name", (-bound_ts, bound_ts), nominal=10.0) # setting nominal also possible
```

Why implement this, rather then the current approach? Because *this would allow most RTC2 implementations to no longer need to work with / import casadi at all.* It would also solve the annoying fact that creating a variable requires editing 3-4 separate parts of an implementation file.

For symmetry, `make_variable()` could work similarly:

```python
self.set_variable("var_name", (-10, 10), nominal = 1.0)
```

**Comment by SGeeversAtVortech** [comment]
In GitLab by @jvande42b on May 6, 2018, 24:46

@vreeken it is a small departure from the method overriding approach to rtc. On the other hand, it is consistent with set_timeseries()

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on May 7, 2018, 17:07

I do not particularly mind adding a convenience function like this if it is a common pattern. I am however afraid it will confuse people about what `path_variables` and `extra_variables` are for. Also, I am not sure how this will tie in to caching of e.g. bounds, which is done at the level of `ModelicaMixin` and `PIMixin`. My guess/worry is that any implementation of `set_path_variable` at that level or deeper will not work, which begs the question of where to implement it.

For the first two lines I would try to keep it in line with the implementation of #1005. That is, use `optimization_problem.variable(...)` with `constant_inputs`.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Aug 28, 2019, 23:02

I think this issue has since been addressed by 802e9c012d13f846df0224e106682e7cfd081bc4. You can just define a new constant input, and refer to it with with `self.variable(...)`.

For scalar values there is no such need, as a parameter already was sufficient.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Aug 28, 2019, 23:02

closed
